index.js

```diff
@@ -108,54 +108,82 @@ module.exports = function prettyFactory (options) {
+    let swapTimeTokens = [
+      { key: 'level' },
+      { delimiter: ' [', requiresAll: ['time'] },
+      { key: 'time' },
+      { delimiter: ']', requiresAll: ['time'] }
+    ]
 
-    const coloredLevel = levels.hasOwnProperty(log.level)
-      ? color[log.level](levels[log.level])
-      : color.default(levels.default)
-    if (opts.levelFirst) {
-      line = `${coloredLevel} ${line}`
+    if (opts.format) {
+      tokens = opts.format
     } else {
-      // If the line is not empty (timestamps are enabled) output it
-      // with a space after it - otherwise output the empty string
-      const lineOrEmpty = line && line + ' '
-      line = `${lineOrEmpty}${coloredLevel}`
-    }
-
-    if (log.name || log.pid || log.hostname) {
-      line += ' ('
-
-      if (log.name) {
-        line += log.name
+      if (opts.levelFirst) {
+        tokens = swapTimeTokens
       }
+      tokens.push({ delimiter: ' (', requiresOne: ['name', 'pid', 'hostname'] })
+      tokens.push({ key: 'name' })
+      tokens.push({ delimiter: '/', requiresAll: ['name', 'pid'] })
+      tokens.push({ key: 'pid' })
+      tokens.push({ delimiter: ' on ', requiresAll: ['hostname'] })
+      tokens.push({ key: 'hostname' })
+      tokens.push({ delimiter: ')', requiresOne: ['name', 'pid', 'hostname'] })
+      tokens.push({ delimiter: ': ' })
+    }
 
-      if (log.name && log.pid) {
-        line += '/' + log.pid
-      } else if (log.pid) {
-        line += log.pid
-      }
+    var line = ''
+    const standardKeys = [
+      'v'
+    ]
 
-      if (log.hostname) {
-        line += ' on ' + log.hostname
+    tokens.forEach((token) => {
+      if (token.key) {
+        standardKeys.push(token.key)
+        switch (token.key) {
+          case 'time':
+            if (log.time) {
+              if (opts.translateTime) {
+                line += formatTime(log.time, opts.translateTime)
+              } else {
+                line += log.time
+              }
+            }
+            break
+          case 'level':
+            const coloredLevel = levels.hasOwnProperty(log.level)
+              ? color[log.level](levels[log.level])
+              : color.default(levels.default)
+            line += coloredLevel
+            break
+          default:
+            if (log.hasOwnProperty(token.key)) {
+              line += log[token.key]
+            }
+        }
+      } else if (token.delimiter) {
+        if (token.requiresOne) {
+          for (var i = 0; i < token.requiresOne.length; i++) {
+            if (log.hasOwnProperty(token.requiresOne[i])) {
+              line += token.delimiter
+              break
+            }
+          }
+        } else if (token.requiresAll) {
+          if (token.requiresAll.every(item => log.hasOwnProperty(item))) {
+            line += token.delimiter
+          }
+        } else {
+          line += token.delimiter
+        }
       }
-
-      line += ')'
-    }
-
-    line += ': '
+    })
```

Comment:
Can you move this formatting logic into its own file under `lib/`?

Reply:
Sure thing, I'll move it over

Reply:
So I started to move this logic to a module in ./lib, but I realized quickly that this effort isn't worth it in my opinion.  There are too many variables that would need to be shared between the two modules that make this more difficult to understand without refactoring most of the codebase and introducing more risk.  I'm sure that's why the original logic was written in index.js as well.  At the end of the day, I've replaced 38 lines ([L120](https://github.com/pinojs/pino-pretty/blob/master/index.js#L120) -L158) of code with 44 lines ([L152](https://github.com/pinojs/pino-pretty/pull/55/files#diff-168726dbe96b3ce427e7fedce31bb0bcR152)-L196).